### ISSUE TITLE
cpp.Pointer is null when build for  windows x64

### ISSUE THREAD

**Issue by sonygod** [HaxeFoundation/hxcpp#773]
cpp.Pointer is null when build for  windows x64

build and test use this project

https://github.com/wighawag/hxcpp-test

and win32 is build and run ok

but  when target for x64,( -D HXCPP_M64 )the pointer is null

	package;

	import engine.Rectangle;
	import cpp.ConstPointer;
	import cpp.Pointer;

	class App {
		public static var instance:AppInstance; // Gets set from your "main" routine

		public static function init(value:Int):Int {
			instance.init();
			return 0;
		}
	

		public static function update(rect:Pointer<Rectangle>):Int {
			rect.ref.set_values(4, 200);
				instance.update(rect.ref.area());

			return 0;
		}
	}

//rect:Pointer<Rectangle> is null when target x64,but ok in win32?

**Comment by tobil4sk** [comment]
The problem is that the native c++ from that codebase defines its own version of ::cpp::Pointer instead of using the actual definition: https://github.com/wighawag/hxcpp-test/blob/40d6c9f05ef27ede1e0ab31ba01272c84b1a2593/cpp/src/haxetest.cpp#L8. I think 32 bit worked by accident, but this isn't correct code. It can be "corrected" for this example by adding the copy constructor, presumably something to do with trivially-copyable types:
```cpp
namespace cpp
{

	template<typename T>
	class Pointer
	{
	public:
		T *ptr;
		inline Pointer(const T *inValue) : ptr((T*)inValue) { }
		inline Pointer(const Pointer &inRHS) : ptr(inRHS.ptr) {}
	};

}
```

However, to avoid other issues, it's better to use the real types instead of redefining incomplete versions:

```diff
diff --git a/cpp/CMakeLists.txt b/cpp/CMakeLists.txt
index f148c20..04d58d7 100644
--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -8,7 +8,11 @@ set(libs
 	../../lib/Main-debug	
    )
 
-include_directories("include")
+include_directories(
+	"include"
+	D:/path/to/hxcpp/include/
+	../lib/include
+)
 
 add_executable(${CMAKE_PROJECT_NAME} ${sources})
 
@@ -16,6 +20,8 @@ target_link_libraries(${CMAKE_PROJECT_NAME} ${libs})
 
 
 if(WIN32)
+   target_compile_definitions(${CMAKE_PROJECT_NAME} PRIVATE -DHX_WINDOWS -DHXCPP_M64)
+
    set_target_properties(${CMAKE_PROJECT_NAME} PROPERTIES LINK_FLAGS_DEBUG "/SUBSYSTEM:CONSOLE")
    set_target_properties(${CMAKE_PROJECT_NAME} PROPERTIES COMPILE_DEFINITIONS_DEBUG "_CONSOLE")
    set_target_properties(${CMAKE_PROJECT_NAME} PROPERTIES LINK_FLAGS_RELWITHDEBINFO "/SUBSYSTEM:CONSOLE")
diff --git a/cpp/src/haxetest.cpp b/cpp/src/haxetest.cpp
index d61d43d..da10a94 100644
--- a/cpp/src/haxetest.cpp
+++ b/cpp/src/haxetest.cpp
@@ -1,25 +1,7 @@
 #include <iostream>
 #include "Rectangle.h"
 
-namespace cpp
-{
-
-	template<typename T>
-	class Pointer
-	{
-	public:
-		T *ptr;
-		inline Pointer(const T *inValue) : ptr((T*)inValue) { }
-	};
-
-}
-
-class App_obj{
-public:
-	static int init(int value);
-	static int update(::cpp::Pointer<Rectangle> rect);
-
-};
+#include <App.h>
 
 
 

```

It should also probably be using `@:nativeGen` on `App.hx`.

So in short, the problem here was user code and not a genuine issue with hxcpp.
